File: 2021/day_7.rs
use std::fs;

fn load_data(filepath:&str) -> String {
  let file = fs::read_to_string(filepath);
  if file.is_ok() {
    return file.unwrap();
  }
  panic!("File not found or unreadable");
  }

fn median(nums: &mut Vec<i32>) -> i32 {
  nums.sort();
  let med = nums.len() / 2;
  return nums[med];
}

fn first_star(data:&String) -> i32 {
  let mut crabs:Vec<i32> = data.split(',').map(|c| c.parse::<i32>().unwrap()).collect();

  let med = median(&mut crabs);

  return crabs.iter().map(|c| (med - c).abs()).sum();
}

fn second_star(data:&String) -> i32 {
  let crabs:Vec<i32> = data.split(',').map(|c| c.parse::<i32>().unwrap()).collect();

  let min = *crabs.iter().min().unwrap();
  let max = *crabs.iter().max().unwrap();
  let mut tot:i32 = crabs.iter().map(|c| {let x = (max - c).abs(); return (x*(x+1))/2}).sum();
  let mut sum:i32;

  for step in min..max {
    sum = crabs.iter().map(|c| {let x = (step - c).abs(); return (x*(x+1))/2}).sum();
    if sum < tot {
      tot = sum;
    }
  }

  return tot;
}

fn main() {
  let data = load_data("input.txt");
  println!("First star: {}", first_star(&data));
  println!("Second star: {}", second_star(&data));
}

File: 2024/day_7.py
def load_input(source):
    with open(source) as f:
        data = [l.strip() for l in f.readlines()]
    return data

def parse_input(data):
    eqs = []
    for r in data:
        val, inps = r.split(': ')
        eqs.append((int(val), [int(i) for i in inps.split()]))
    return eqs

def is_valid(val, inps, concat=False):
    chunks = [inps[0]]
    for i in inps[1:]:
        n_chunks = []
        for c in chunks:
            n_chunks.append(c + i)
            n_chunks.append(c * i)
            if concat:
                n_chunks.append(int(str(c) + str(i)))
        chunks = n_chunks
    if val in chunks:
        return True
    return False


def first_star(data):
    tot = 0
    for r in data:
        if is_valid(r[0], r[1]):
            tot += r[0]
    return tot

def second_star(data):
    tot = 0
    for r in data:
        if is_valid(r[0], r[1], concat=True):
            tot += r[0]
    return tot


def solution(source):
    data = parse_input(load_input(source))
    print("Day 7")
    print("First star:", str(first_star(data)))
    print("Second star:", str(second_star(data)))
    print("-------------------------------------")


if __name__ == "__main__":
    solution("input.txt")
